src/lib/io/sol-uart-impl-riot.c

```diff
@@ -119,48 +140,119 @@ sol_uart_open(const char *port_name, const struct sol_uart_config *config)
-        uart_tx_dispatch(uart, -1);
+    SOL_PTR_VECTOR_FOREACH_IDX (&uart->pending_blobs, blob, i) {
+        uart_tx_dispatch(uart, blob, -ECANCELED);
+        sol_blob_unref(blob);
     }
 
-    if (uart->async.handler)
-        sol_interrupt_scheduler_uart_stop(uart->id, uart->async.handler);
-    uart_poweroff(uart->id);
+    if (uart->tx_writer)
+        sol_timeout_del(uart->tx_writer);
+
+    if (uart->rx_reader)
+        sol_timeout_del(uart->rx_reader);
 
+    if (uart->handler)
+        sol_interrupt_scheduler_uart_stop(uart->id, uart->handler);
+
+    uart_poweroff(uart->id);
+    sol_ptr_vector_clear(&uart->pending_blobs);
+    sol_buffer_fini(&uart->rx_buffer);
     free(uart);
 }
 
-SOL_API bool
-sol_uart_write(struct sol_uart *uart, const uint8_t *tx, unsigned int length, void (*tx_cb)(void *data, struct sol_uart *uart, uint8_t *tx, int status), const void *data)
+SOL_API int
+sol_uart_write(struct sol_uart *uart, struct sol_blob *blob)
 {
-    SOL_NULL_CHECK(uart, false);
-    SOL_EXP_CHECK(uart->async.tx_buffer, false);
+    bool created_writer = false;
+    int r = 0;
+
+    SOL_NULL_CHECK(uart, -EINVAL);
+    SOL_NULL_CHECK(blob, -EINVAL);
+
+    if (!uart->tx_writer) {
+        uart->tx_writer = sol_timeout_add(0, uart_tx_cb, uart);
+        SOL_NULL_CHECK(uart->tx_writer, -ENOMEM);
+        created_writer = true;
+    }
 
-    uart->async.tx_buffer = tx;
-    uart->async.tx_cb = tx_cb;
-    uart->async.tx_user_data = data;
-    uart->async.tx_length = length;
 
-    uart->async.tx_writer = sol_timeout_add(0, uart_tx_cb, uart);
+    r = -EOVERFLOW;
+    SOL_NULL_CHECK_GOTO(sol_blob_ref(blob), err_blob);
+    r = sol_ptr_vector_append(&uart->pending_blobs, blob);
+    SOL_INT_CHECK_GOTO(r, < 0, err_append);
+    return 0;
+
+err_append:
+    sol_blob_unref(blob);
+err_blob:
+    if (created_writer) {
+        sol_timeout_del(uart->tx_writer);
+        uart->tx_writer = NULL;
+    }
+    return r;
+}
+
+SOL_API int
+sol_uart_read(struct sol_uart *uart, struct sol_buffer *buf)
+{
+    enum sol_buffer_flags flags;
+    size_t min;
+    int r;
+
+    SOL_NULL_CHECK(uart, -EINVAL);
+    SOL_NULL_CHECK(buf, -EINVAL);
+
+    if (!uart->rx_buffer.used)
+        return 0;
+
+    if (buf->flags & SOL_BUFFER_FLAGS_FIXED_CAPACITY) {
+        min = sol_min(uart->rx_buffer.used, (buf->capacity - buf->used));
+        if (!min)
+            return -ENOMEM;
+        r = sol_buffer_append_bytes(buf, uart->rx_buffer.data, min);
+        SOL_INT_CHECK(r, < 0, r);
+        return sol_buffer_remove_data(&uart->rx_buffer, 0, min);
+    }
+
+    //Swap buffers
+    flags = buf->flags;
+
+    memcpy(buf, &uart->rx_buffer, sizeof(struct sol_buffer));
+
+    sol_buffer_init_flags(&uart->rx_buffer, NULL, 0,
+        SOL_BUFFER_FLAGS_DEFAULT | SOL_BUFFER_FLAGS_NO_NUL_BYTE);
+    buf->flags = flags;
+
+    if (!(buf->flags & SOL_BUFFER_FLAGS_NO_NUL_BYTE)) {
+        r = sol_buffer_ensure_nul_byte(buf);
+        SOL_INT_CHECK(r, < 0, r);
+    }
```

Comment:
@barbieri Thinking about this one again. I don't think this is the right solution. We should append the whole buffer here and not swap it. If one is using the `buf` to accumulate all the received data, the swap will override the previous read data.

I think we should change the code to
```c
if (buf->flags & SOL_BUFFER_FLAGS_FIXED_CAPACITY) { 
//Leave as is.
}
//The new code
sol_buffer_append_buffer(buf, &uart->rx_buffer);
sol_buffer_reset(&uart->rx_buffer);
return 0;

```